multicursor format kluge for sort-ns-references

pbwolf opened the pull request:
## What has changed?

cljfmt sorting of namespace references is restored, but the cursor may wind up in an unexpected place if it does so.

The issue title was "Format on save + cljfmt fails..." but the particular problem was that multicursor-capable formatting relied on cljfmt to change only whitespace.

The multicursor-capable formatter used "console.error" when that condition was violated, but the messages may not have been noticed.

This PR makes the multicursor-capable formatter fall back to replacing the whole range if it cannot calculate micro-edits of whitespace. Whole-range replacement will be unkind to cursors that were within the range.

Edits 2025-04-27: The PR is revised to refine the kluge slightly: using the same technique as for cursor-contexts, it locates a `(ns...)` form, and if there is one, then it formats that form separately from the remainder of the document. In that way, each of the two parts produces efficient spaces-only edits that allow cursors to float, or degenerates to making a whole replacement that may shove cursors to the end; but an `(ns...)` form that gets sorted does not prevent cursors after that form from floating.

A test case is added for the "respacer" module for the case in which it stuffs the remainder of the substantively-differing document into a single edit.
 
The CHANGELOG wishfully renames the issue for clarity, saying, (Fixes) "Reformat may do nothing if cljfmt's :sort-ns-references? is true".

Addresses #2789

## My Calva PR Checklist

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Made sure there is an issue registered with a clear problem statement that this PR addresses, (created the issue if it was not present).
    - [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [ ] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [ ] Added to or updated docs in this branch, if appropriate
- [ ] Tests
  - [x] Tested the particular change
  - [ ] Figured if the change might have some side effects and tested those as well.
- [x] Formatted all JavaScript and TypeScript code that was changed. (use the [prettier extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) or run `npm run prettier-format`)
- [x] Confirmed that there are no linter warnings or errors (use the [eslint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint), run `npm run eslint` before creating your PR, or run `npm run eslint-watch` to eslint as you go).

Ping @pez, @bpringe, @corasaurus-hex, @Cyrik

pbwolf committed:
When cljfmt changes more than whitespace, fall back to replacing the entire formatted range

PEZ commented:
> Perhaps, when Calva invokes cljfmt, it should trample `:sort-ns-references?` to false?

It should be considered. This is a tricky one. Depends on if we think we can fix the underlying problem. (Which seems hard?). If the situation is temporary displacing cursors may be an annoyance we can live with. If it is permanent, we may be at a point where we need to decide not to sopport sorting from the formatter?

cc: @bpringe

pbwolf committed:
Merge branch 'dev' into 2789-format-sort-ns-references

pbwolf committed:
Format (ns) and remainder separately to balance multicursor & sort-ns-references

To nudge all cursors with the tides of reformatting, micro edits change only
whitespace. Computing edits is linear-time if we know only whitespace
has changed. But there is 1 case where cljfmt changes text: sort-ns-references.
To balance both multicursor nudging and sort-ns-references,
we format an ns form separately, then append the ordinarily-spaces-only
edits for the rest of the document. As a result, cursors located within
a (ns...) form might move to the end of the form, but other cursors
throughout the document should continue to float.

pbwolf committed:
Merge dev

PEZ commented on CHANGELOG.md:
```diff
@@ -4,9 +4,12 @@ Changes to Calva.
 ## [2.0.503] - 2025-04-27
 
 - Fix: [Release 2.0.502 repl connection fails, breaking Jack-In](https://github.com/BetterThanTomorrow/calva/issues/2799) (Hopefully)
+>>>>>>> dev
```
```suggestion
```

PEZ committed:
Remove diff marker

bpringe commented:
What's the state of this now? Do we only not support ns form sorting in certain situations, or do we never support it?

My opinion is that we should try to support all cljfmt options in all situations, even if that causes some other (at least tolerable) issue. If we decide not to support sorting in certain situations, then I think maybe informing the user in docs would be good and also maybe informing them with an editor popup when the non-supporting case occurs during formatting that allows them to select to not show it again.

pbwolf commented:
> What's the state of this now? Do we only not support ns form sorting in certain situations, or ...?

Calva supports cljfmt's namespace sorting.
